pennylane/interfaces/autograd.py

```diff
@@ -195,18 +195,73 @@ def vjp(ans, self, params, device):  # pylint: disable=unused-argument
             gradient output vector, and computes the vector-Jacobian product
         """
 
-        def gradient_product(g):
-            # In autograd, the forward pass is always performed prior to the backwards
-            # pass, so we do not need to re-unwrap the parameters.
+        saved_grad_matrices = {}
+
+        def _evaluate_grad_matrix(p, grad_matrix_fn):
+            """Convenience function for generating gradient matrices
+            for the given parameter values.
+
+            This function serves two purposes:
+
+            * Avoids duplicating logic surrounding parameter unwrapping/wrapping
+
+            * Takes advantage of closure, to cache computed gradient matrices via
+              the ``saved_grad_matrices`` attribute, to avoid gradient matrices being
+              computed multiple redundant times.
+
+              This is particularly useful when differentiating vector-valued QNodes.
+              Because Autograd requests the vector-grad matrix product,
+              and *not* the full grad matrix, differentiating vector-valued
+              functions will result in multiple backward passes.

```

Comment:
List the arguments here?

Would be good to clarify `p` is the parameters and `grad_matrix_fn` is a string that is either `"jacobian"` or `"hessian"`.

Reply:
```suggestion
              functions will result in multiple backward passes.

            Args:
                p (Sequence): quantum tape parameter values use to evaluate the gradient matrix
                grad_matrix_fn (str): Name of the gradient matrix function. Should correspond to an existing
                    tape method. Currently allowed values include ``"jacobian"`` and ``"hessian"``.
                    
                Returns:
                    array[float]: the gradient matrix
```